src/server/nuclearnet/web_socket_proxy_nuclearnet_server.ts

```diff
@@ -200,22 +203,52 @@ class PacketProcessor {
   }
 
   private maybeSendNextPacket() {
-    if (this.outgoingPackets < this.outgoingLimit) {
-      const next = this.queue.pop()
-      if (next) {
+    const next = this.queue.pop()
+    if (next) {
+      const { event, packet } = next
+      const key = `${event}:${packet.peer.name}:${packet.peer.address}:${packet.peer.port}`
+      if (this.canSendEvent(key)) {
         const { event, packet } = next
         let isDone = false
-        const done = () => {
+        const stats = this.getStatsForEvent(key)
+        stats.active++
+        const start = this.clock.performanceNow()
+        const done = (processingTime?: number) => {
           if (!isDone) {
-            this.outgoingPackets--
+            // Update our performance tracking information
+            stats.active -= 1
+            if (processingTime != null) {
+              // Calculate network latency
+              const latency = this.clock.performanceNow() - start
+              const n = stats.active
+              stats.latency = (stats.latency * n + latency) / (n + 1)
+              stats.processingTime = (stats.processingTime * n + processingTime) / (n + 1)
+            }
             isDone = true
             this.maybeSendNextPacket()
           }
         }
-        this.outgoingPackets++
-        this.socket.volatileSend(event, packet, done)
+        // this.socket.volatileSend(event, packet, done)
+        this.socket.send(event, packet, done)
+        stats.lastSent = this.clock.performanceNow()
         this.clock.setTimeout(done, this.timeout)
       }
     }
   }
+
+  private canSendEvent(event: string): boolean {
+    const stats = this.getStatsForEvent(event)
+    const maxRate = stats.processingTime / stats.latency
```

Comment:
I think if we are sending based on a rate, we can just use processing time alone. We want to send packets at a rate that matches the processing time on the other end